### PR TITLE
consume as a broccoli tree instead of re-creating EmberApp

### PR DESCRIPTION
**Pull request by fivetanley**
This should fix an issue where babel plugins get re-registered twice.
There's a myriad of situations where copying over the options from an
existing application could result in misconfigurd or duplicate builds.

Instead of ripping `options` off EmberApp, we use the function exported
by the Ember CLI application's `ember-cli-build.js` file to return a
broccli tree. This allows the app to use any broccoli plugins (without
creating an addon) and also re-runs any initialization steps we might
have missed by creating the `EmberApp` ourselves. This also eliminates
reliance on the private `javascript()` method on `EmberApp`.

### CODE REVIEW

**Comment by fivetanley** [comment]
I'm testing this against the ember-api-docs now to make sure extra files aren't included in `fastboot` folder, I'll post back shortly

**Comment by fivetanley** [comment]
One downside to this approach is that we end up building css/etc twice when we only want to build JS twice. I think we can configure this with options:

https://github.com/ember-cli/ember-cli/blob/65f6ffafc9f53f1ced3ea9563ef2b27c40763763/lib/broccoli/ember-app.js#L220

**Comment by fivetanley** [comment]
Some of the project directives like the `appFile` field are missing in package.json. I'm looking into that.

**Comment by jasonmit** [comment]
I was trying this branch out because it touched some similar logic I was working on to fix #141.  However, it looks like production builds are failing I suspect due to the changes around asset map.

```sh
ember fastboot --environment production --port=1234       ⏎ ✖ ✚ ✭
version: 2.4.1
Built project successfully. Stored in "fastboot-dist".
Installing FastBoot npm dependencies
Error: ENOENT: no such file or directory, open 'fastboot-dist/fastboot/asset-rev-example.js'
    at Error (native)
    at Object.fs.openSync (fs.js:584:18)
    at Object.fs.readFileSync (fs.js:431:33)
    at new EmberApp (/home/elek/projects/ember-cli-fastboot/node_modules/ember-fastboot-server/lib/ember-app.js:39:20)
    at new FastBootServer (/home/elek/projects/ember-cli-fastboot/node_modules/ember-fastboot-server/lib/server.js:33:14)
    at /home/elek/projects/ember-cli-fastboot/lib/commands/fastboot.js:32:22
    at lib$rsvp$$internal$$tryCatch (/home/elek/projects/ember-cli-fastboot/node_modules/rsvp/dist/rsvp.js:1036:16)
    at lib$rsvp$$internal$$invokeCallback (/home/elek/projects/ember-cli-fastboot/node_modules/rsvp/dist/rsvp.js:1048:17)
    at lib$rsvp$$internal$$publish (/home/elek/projects/ember-cli-fastboot/node_modules/rsvp/dist/rsvp.js:1019:11)
    at lib$rsvp$asap$$flush (/home/elek/projects/ember-cli-fastboot/node_modules/rsvp/dist/rsvp.js:1198:9)
    at doNTCallback0 (node.js:430:9)
    at process._tickCallback (node.js:359:13)
```

**Comment by fivetanley** [comment]
@jasonmit I figured out the source of that (which was also the source of my test failure). Due to this change, you'll need to enable `generateAssetMap: true` in the `fingerprint` options passed to `new EmberApp` in your Brocfile.

**Comment by fivetanley** [comment]
I need to figure out a way to cancel the build if this isn't present...

**Comment by fivetanley** [comment]
An example: https://github.com/fivetanley/ember-cli-fastboot/blob/require-ember-app/tests/fixtures/module-whitelist/ember-cli-build.js#L5

File: lib/broccoli/fastboot-config.js [hunk 1]
```diff
@@ -82,7 +83,6 @@ FastBootConfig.prototype.readAssetManifest = function() {
 
   try {
     var assetMap = JSON.parse(fs.readFileSync(assetMapPath));
-    fs.unlinkSync(assetMapPath);
```
Comment (fivetanley) [inline lib/broccoli/fastboot-config.js]: this change is no longer needed, i was trying to debug something.

**Comment by danmcclain** [comment]
I can verify that this works, I've been using this PR for deploying dockyard.com

**Comment by sepowitz** [comment]
hi all - any reason this hasn't been merged yet?  happy to help!

**Comment by tomdale** [comment]
@sepowitz Just my own laziness. :) I am reviewing it now and hope to have it merged soon.

**Comment by hhff** [comment]
:+1:

**Comment by sepowitz** [comment]
@tomdale right on!

**Comment by tomdale** [comment]
@fivetanley So this looks great. I started rebasing this onto master, and I think there may be a regression in this refactor in that it looks like output paths are not respected. I pushed the rebased version to the `fivetanley-require-ember-app` branch; can you take a look? It has one failing test, which is that the manifest is using the `outputPaths` but it is not finding anything there.

**Comment by hhff** [comment]
ya I noticed that @fivetanley's branch is against 0.6.0, but we're at 0.6.2

**Pull request merged**